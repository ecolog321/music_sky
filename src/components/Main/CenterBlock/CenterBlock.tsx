"use client";

import Image from "next/image";
import styles from "./centerblock.module.css";
import sharedStyles from "../../shared.module.css";
import { TrackList } from "../TrackList/TrackList";
import clsx from "clsx";
import { TrackType } from "../../../types/types";
import { Filter } from "@components/Filter/Filter";
import { filterFresh } from "@components/Filter/Filter.data";
import { ChangeEvent, FC, useCallback, useEffect, useState } from "react";
import { useAppDispatch, useAppSelector } from "../../../hooks/store";
import { setFilters, setPlaylist } from "../../../store/features/playlistSlise";
import { saveUser } from "../../../store/features/authSlice";
import { SearchField } from "./SearchFeild/SearchFeild";
import { createUniqueArray } from "../../../utils/uniqueArray";

type Props = {
  tracks: TrackType[];
};

export const CenterBlock: FC<Props> = ({ tracks }) => {
  const [filterValue, setFilterValue] = useState<string | null>(null);
  const [authors, setAuthors] = useState<string[]>([]);
  const [release, setRelease] = useState<string[]>([]);
  const [genre, setGenre] = useState<string[]>([]);
  const [currentFilter, setCurrentFilter] = useState<string[]>([]);
  const [searchField, setSearchField] = useState("");
  const dispatch = useAppDispatch();
  const filteredTracks = useAppSelector(
    (store) => store.playlist.filteredPlaylist
  );

  const filterData = [
    {
      title: "исполнителю",
      value: "author",
      selected: useAppSelector((store) => store.playlist.filterOptions.author),
    },
    {
      title: "году выпуска",
      value: "release",
      selected: [useAppSelector((store) => store.playlist.filterOptions.order)],
    },
    {
      title: "жанру",
      value: "genre",
      selected: useAppSelector((store) => store.playlist.filterOptions.genre),
    },
  ];

  const handleSearchField = useCallback((e:ChangeEvent<HTMLInputElement>)=> setSearchField(e.target.value),[])

  const toogleAuthors = () => createUniqueArray(tracks, "author");

  const toogleGenre = () => createUniqueArray(tracks, "genre");

  useEffect(() => {
    setAuthors(toogleAuthors());
    setGenre(toogleGenre());
    setRelease(filterFresh);
  }, [tracks]);

  useEffect(() => {
    dispatch(setPlaylist({ tracks }));
  }, [dispatch, tracks]);

  useEffect(() => {
    dispatch(setFilters({ searchField }));
  }, [searchField]);

  useEffect(() => {
    dispatch(
      saveUser([
        localStorage.getItem("email") || "",
        localStorage.getItem("access") || "",
        localStorage.getItem("refresh") || "",
      ])
    );
  }, [dispatch]);

  const changeFilterValue = (value: string) => {
    setFilterValue((prev) => (prev === value ? null : value));

    switch (value) {
      case "author":
        setCurrentFilter(authors);
        break;
      case "genre":
        setCurrentFilter(genre);
        break;
      case "release":
        setCurrentFilter(release);
        break;

      default:
        break;
    }
  };

  return (
    <div className={styles.main__centerblock}>
      <SearchField handleSearchField={handleSearchField} />
      <h2 className={styles.centerblock__h2}>Треки</h2>
      <div className={styles.centerblock__filter}>
        <div className={sharedStyles.filter__title}>Искать по:</div>
        {filterData.map((item) => (
          <Filter
            isOpen={filterValue === item.value}
            value={item.value}
            title={item.title}
            list={currentFilter}
            key={item.value}
            onClick={changeFilterValue}
            selected={item.selected}
          />
        ))}
      </div>
      <div className={styles.centerblock__content}>
        <div className={styles.content__title}>
          <div className={clsx(styles.playlist_title__col, sharedStyles.col01)}>
            Трек
          </div>
          <div className={clsx(styles.playlist_title__col, sharedStyles.col02)}>
            Исполнитель
          </div>
          <div className={clsx(styles.playlist_title__col, sharedStyles.col03)}>
            Альбом
          </div>
          <div className={clsx(styles.playlist_title__col, sharedStyles.col04)}>
            <Image
              src="/img/icons/watch.svg"
              height={12}
              width={14}
              alt="like"
            ></Image>
          </div>
        </div>
        <TrackList tracks={filteredTracks} />
      </div>
    </div>
  );
};
