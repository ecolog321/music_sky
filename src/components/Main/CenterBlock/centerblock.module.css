.main__centerblock {
    width: auto;
    -webkit-box-flex: 3;
    -ms-flex-positive: 3;
    flex-grow: 3;
    padding: 20px 40px 20px 111px;
  }


  
  .centerblock__h2 {
    font-style: normal;
    font-weight: 400;
    font-size: 64px;
    line-height: 72px;
    letter-spacing: -0.8px;
    margin-bottom: 45px;
  }
  
  .centerblock__filter {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    margin-bottom: 51px;
    gap: 10px;
  }
  
  .centerblock__content {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: vertical;
    -webkit-box-direction: normal;
    -ms-flex-direction: column;
    flex-direction: column;
  }

 


  .content__title {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: start;
    -webkit-box-pack: justify;
    -ms-flex-pack: justify;
    justify-content: space-between;
    margin-bottom: 24px;
  }

  
  