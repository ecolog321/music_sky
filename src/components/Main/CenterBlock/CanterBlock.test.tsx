import renderer from "react-test-renderer";
import ReduxProvider from "../../../store/ReduxProvider";
import { CenterBlock } from "./CenterBlock";

describe("TrackComponent", () => {
  it("render track", () => {
    const tree = renderer
      .create(
        <ReduxProvider>
          <CenterBlock tracks={[]} />
        </ReduxProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
