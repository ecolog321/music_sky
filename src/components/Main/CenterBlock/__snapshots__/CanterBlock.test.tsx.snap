// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackComponent render track 1`] = `
<div
  className="main__centerblock"
>
  <div
    className="centerblock__search"
  >
    <svg
      className="search__svg"
    >
      <use
        xlinkHref="/img/icons/sprite.svg#icon-search"
      />
    </svg>
    <input
      className="search__text"
      name="search"
      onChange={[Function]}
      placeholder="Поиск"
      type="search"
    />
  </div>
  <h2
    className="centerblock__h2"
  >
    Треки
  </h2>
  <div
    className="centerblock__filter"
  >
    <div
      className="filter__title"
    >
      Искать по:
    </div>
    <div
      className="filter__block"
    >
      <button
        className="filter__button"
        onClick={[Function]}
      >
        исполнителю
      </button>
    </div>
    <div
      className="filter__block"
    >
      <button
        className="filter__button"
        onClick={[Function]}
      >
        году выпуска
      </button>
    </div>
    <div
      className="filter__block"
    >
      <button
        className="filter__button"
        onClick={[Function]}
      >
        жанру
      </button>
    </div>
  </div>
  <div
    className="centerblock__content"
  >
    <div
      className="content__title"
    >
      <div
        className="playlist_title__col col01"
      >
        Трек
      </div>
      <div
        className="playlist_title__col col02"
      >
        Исполнитель
      </div>
      <div
        className="playlist_title__col col03"
      >
        Альбом
      </div>
      <div
        className="playlist_title__col col04"
      >
        <img
          alt="like"
          data-nimg="1"
          decoding="async"
          height={12}
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="/img/icons/watch.svg"
          style={
            {
              "color": "transparent",
            }
          }
          width={14}
        />
      </div>
    </div>
    <div
      className="content__playlist"
    >
      <h2>
        Треки не найдены
      </h2>
    </div>
  </div>
</div>
`;
